### PR TITLE
Feature/17 data preprocessing pipeline

### LINKED ISSUE #17: Data preprocessing pipeline: wczytywanie z HF Hub, zapisywanie do HF Hub, separacja logów i dodanie README
# Wczytywanie z HF Hub, separacja logów, dodanie README oraz refaktoryzacja konfiguracji do dataclass

##  Problem
Obecny pipeline zapisuje do głównego datasetu niepotrzebne dane diagnostyczne (np. `kept`, `filter_reason`), co zanieczyszcza strukturę. Brakuje możliwości wczytywania danych prosto z Hugging Face Hub oraz dokumentacji projektu. Dodatkowo zarządzanie parametrami wejściowymi jest rozproszone i mało elastyczne, co utrudnia utrzymanie kodu.

##  Rozwiązanie
1. **Refaktoryzacja konfiguracji:** Przejście na strukturę `dataclass`, aby uporządkować parametry. Wprowadzenie jednego ujednoliconego pola (np. `input_source`), które inteligentnie obsłuży zarówno lokalne ścieżki, jak i nazwy repozytoriów.
2. **Wczytywanie z HF Hub:** Bezpośrednie pobieranie datasetu ze zunifikowanego źródła.
3. **Zapisywanie do HF Hub:** Pipeline aktualnie obsługuje zapisywanie zbioru danych lokalnie w formacie datasets, ale brakowało opcji pushowania – zostanie ona dodana. Zostanie również dodany podział datasetu na części train i val.
4. **Podział zapisu:** - **Czysty dataset:** Zawiera tylko niezbędne pola (np. `source_url`, `question`, `solution`).
   - **Logi lokalne:** Osobny plik JSONL ze wszystkimi metadanymi i wynikami filtrowania do celów diagnostycznych.
5. **Dokumentacja:** Utworzenie pliku `README.md`.

##  Zadania
- [x] Refaktoryzacja zarządzania konfiguracją do `dataclass` (wprowadzenie `input_source`).
- [x] Implementacja pobierania danych prosto z HF Hub (oraz bezpieczna obsługa stringów dla plików lokalnych/zdalnych).
- [x] Dodanie opcji wysyłania (push) czystego zbioru danych na HF Hub z podziałem na train i val.
- [x] Utworzenie logiki generującej czysty rekord oraz pełny rekord z logami.
- [x] Zaimplementowanie zapisu czystego datasetu oraz osobnego pliku z logami lokalnymi.
- [x] Dodanie `README.md` z instrukcją instalacji, ustawiania konfiguracji i uruchomienia pipeline'u.

### PR DESCRIPTION
**Pull request by erybie222**
Zmiany są opisane w issue.

### CODE REVIEW

File: data-preprocessing/models.py [hunk 1]
```diff
@@ -28,25 +36,28 @@ def __init__(self, input_file: Path, output_file: Path, dataset_dir: Path, check
...
+            self.raw_data.extend(loaded_records)
+            self.stats["loaded"] += len(loaded_records)
 
-                except json.JSONDecodeError as e:
-                    print(f"Warning: JSON parsing error at line {line_number}. Skipping this record. Details: {e}")
+            debug("Loading data", f"Successfully loaded {len(loaded_records)} records.")
 
-        debug("Loading data", f"Successfully loaded {self.stats['loaded']} records.")
+        except Exception as e:
+            print(f"Error: Failed to load dataset from {self.input_source}. Details: {e}")
```
Comment (Igor-S-666) [inline data-preprocessing/models.py]: I think there should be raise instead of print, as raise will halt execution of program, while print will not

File: data-preprocessing/models.py [hunk 2]
```diff
@@ -344,31 +384,46 @@ def run(self) -> None:
...
                                     if not self.quiet:
                                         print("    -> Cleaning answer LaTeX...")
-                                    record["raw_answer"] = self._fix_latex_solution(normalize_latex(raw_answer))
-                                    record["answer_post_index"] = answer_post_idx
+                                    raw_answer_clean = self._fix_latex_solution(normalize_latex(raw_answer))
 
                                     if not self.quiet:
                                         print("    -> Rewriting answer...")
-                                    solution = self._rewrite_answer(question_clean, raw_answer)
+                                    rewritten = self._rewrite_answer(question_clean, raw_answer)
```
Comment (Igor-S-666) [inline data-preprocessing/models.py]: It seems like there should be raw_answer_clean not raw_answer

File: data-preprocessing/utils.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+) -> None:
+    """Saves HuggingFace dataset locally (Path) or pushes to HF Hub (str repo id), with train/val split."""
+    input_path = Path(input_file_path)
+    if not input_path.exists() or input_path.stat().st_size == 0:
+        print(f"Skipping dataset save: {input_path} is empty or does not exist.")
+        return
+
+    raw = load_dataset("json", data_files=str(input_path), split="train")
+    splits = raw.train_test_split(test_size=val_size, seed=seed)
+    dataset = DatasetDict({"train": splits["train"], "val": splits["test"]})
```
Comment (Igor-S-666) [inline data-preprocessing/utils.py]: According to huggingface, it should be "validation", not "val" (https://huggingface.co/docs/datasets/loading)
Also, AI suggests it may cause some errors during loading

File: data-preprocessing/config.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    # INPUT_SOURCE and DATASET_DESTINATION can be either a local file or a Hugging Face Hub dataset, depending on LOAD_FROM_HUB and PUSH_TO_HUB flags
+    INPUT_SOURCE: str | Path = field(init=False)
+    DATASET_DESTINATION: str | Path = field(init=False)
+
+    def __post_init__(self) -> None:
+        self.INPUT_FILE = self.DATA_DIR / "input" / "forum_example_fixed.jsonl"
+        self.OUTPUT_FILE = self.DATA_DIR / "output" / "pipeline_output.jsonl"
+        self.LOG_FILE = self.DATA_DIR / "logs" / "pipeline_logs.jsonl" if self.SAVE_LOGS else None
+        self.DATASET_FILE = self.DATA_DIR / "dataset" / "pipeline_output.jsonl"
```
Comment (Igor-S-666) [inline data-preprocessing/config.py]: It is not used outside of this 1 function, so it is better to make it local variable

File: data-preprocessing/models.py [hunk 5]
```diff
@@ -284,32 +333,26 @@ def run(self) -> None:
...
                     if not filt["keep"]:
                         if not self.quiet:
                             print(f"    -> DISCARDED: {filt['reason']}")
                         self.stats["filtered_out"] += 1
-                        out_f.write(json.dumps(record, ensure_ascii=False) + "\n")
                     else:
                         try:
                             if not self.quiet:
                                 print("    -> KEPT. Classifying question...")
                             classification = self._classify_question(question)
```
Comment (Igor-S-666) [inline data-preprocessing/models.py]: shouldn't it use clean_question instead of normal question?

File: data-preprocessing/utils.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        "Authorization": f"Bearer {config.API_KEY}",
+        "Content-Type": "application/json",
+    }
+    payload = {
+        "model": config.MODEL,
+        "messages": [
+            {"role": "system", "content": system_prompt},
+            {"role": "user", "content": user_prompt},
+        ],
+        "temperature": 0.2,
```
Comment (Igor-S-666) [inline data-preprocessing/utils.py]: It seems to be a magic number; preferably move that to config.py

File: data-preprocessing/utils.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    }
+    response = requests.post(
+        config.BASE_URL,
+        headers=headers,
+        json=payload,
+        timeout=60,
```
Comment (Igor-S-666) [inline data-preprocessing/utils.py]: Reasonable default, yet again magic number; move it to config.py

**Review by Igor-S-666 — COMMENTED** [review]
generally good, yet I found few small issues

File: data-preprocessing/models.py (hunk 1 above)
Comment (erybie222) [reply data-preprocessing/models.py]: You are right. I changed thtat to raise Runtime error, because it could be either file not found error or error related to loading from hub , which could be various.

File: data-preprocessing/models.py (hunk 2 above)
Comment (erybie222) [reply data-preprocessing/models.py]: Yes, I fixed that.

File: data-preprocessing/utils.py (hunk 3 above)
Comment (erybie222) [reply data-preprocessing/utils.py]: Right, I renamed that.

File: data-preprocessing/config.py (hunk 4 above)
Comment (erybie222) [reply data-preprocessing/config.py]: do you mean about DATA_DIR ,SAVE_LOGS, PUSH_TO_HUB AND LOAD_FROM_HUB? ( I changed them to local)

Comment (erybie222) [reply data-preprocessing/config.py]: rest of them are used in pipeline.py

File: data-preprocessing/models.py (hunk 5 above)
Comment (erybie222) [reply data-preprocessing/models.py]: Yes, it should. I fixed that.

File: data-preprocessing/utils.py (hunk 6 above)
Comment (erybie222) [reply data-preprocessing/utils.py]: Ok, moved.

File: data-preprocessing/utils.py (hunk 7 above)
Comment (erybie222) [reply data-preprocessing/utils.py]: Ok, moved.

Comment (erybie222) [reply data-preprocessing/utils.py]: Moved seed for  splitting dataset as well
